File: api/store.py
"""In-memory indexes over the real dictionary.json files."""

from __future__ import annotations

import copy
import json
import unicodedata
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

from api import REPO_ROOT
from api.catalog import DatasetRef, load_catalog
from api.settings import API_ORIGIN

APOSTROPHES = dict.fromkeys(map(ord, "‘’ʼ´`"), "'")
META_SKIP = frozenset({"entries"})


def fold(text: str) -> str:
    text = unicodedata.normalize("NFC", text).strip().casefold()
    text = text.translate(APOSTROPHES)
    return " ".join(text.split())


def fold_loose(text: str) -> str:
    decomposed = unicodedata.normalize("NFD", fold(text))
    return "".join(ch for ch in decomposed if unicodedata.category(ch) != "Mn")


def _headword_keys(entry: Dict[str, Any]) -> List[str]:
    values = [entry.get("headword") or ""]
    for variant in entry.get("variants") or []:
        if isinstance(variant, str):
            values.append(variant)
        elif isinstance(variant, dict):
            values.append(variant.get("form") or variant.get("headword") or "")
    return [value for value in values if value]


@dataclass
class Dataset:
    ref: DatasetRef
    document: Dict[str, Any]
    entries: List[Dict[str, Any]] = field(default_factory=list)
    by_id: Dict[str, Dict[str, Any]] = field(default_factory=dict)
    by_fold: Dict[str, List[Dict[str, Any]]] = field(default_factory=dict)
    by_loose: Dict[str, List[Dict[str, Any]]] = field(default_factory=dict)

    @property
    def kind(self) -> str:
        return self.ref.kind

    @property
    def missing_term(self) -> Dict[str, Any]:
        return dict(self.document.get("missing_term") or {
            "status": "error",
            "code": "TERM_NOT_FOUND",
            "message": "Translation not available in the verified database.",
        })

    def metadata(self) -> Dict[str, Any]:
        payload = {
            key: copy.deepcopy(value)
            for key, value in self.document.items()
            if key not in META_SKIP
        }
        payload["dataset"] = self.ref.key
        payload["kind"] = self.ref.kind
        if self.ref.canonical_key:
            payload["canonical_dataset"] = self.ref.canonical_key
        payload["entry_count"] = len(self.entries)
        return payload


class Store:
    def __init__(self, datasets: Dict[str, Dataset]) -> None:
        self.datasets = datasets

    @classmethod
    def load(cls) -> "Store":
        catalog = load_catalog()
        loaded: Dict[str, Dataset] = {}
        for key, ref in catalog.items():
            with ref.json_path.open(encoding="utf-8") as handle:
                document = json.load(handle)
            dataset = Dataset(ref=ref, document=document)
            if ref.kind == "lexicon":
                _index_entries(dataset, document.get("entries") or [])
            loaded[key] = dataset

        angola = loaded["angola"]
        canonical = loaded[angola.ref.canonical_key]
        angola.entries = canonical.entries
        angola.by_id = canonical.by_id
        angola.by_fold = canonical.by_fold
        angola.by_loose = canonical.by_loose
        return cls(loaded)

    def get(self, key: str) -> Optional[Dataset]:
        return self.datasets.get(key)

    def require(self, key: str) -> Dataset:
        dataset = self.get(key)
        if dataset is None:
            raise KeyError(key)
        return dataset

    def resolve_key(self, family: str, variety: Optional[str] = None) -> Optional[str]:
        if variety:
            return f"{family}/{variety}"
        return family

    def catalog(self) -> List[Dict[str, Any]]:
        items = []
        for key in sorted(self.datasets):
            dataset = self.datasets[key]
            meta = dataset.metadata()
            items.append(
                {
                    "dataset": key,
                    "kind": dataset.kind,
                    "language": meta.get("language"),
                    "language_name": meta.get("language_name"),
                    "iso_639_3": meta.get("iso_639_3"),
                    "entry_count": meta.get("entry_count"),
                    "canonical_dataset": meta.get("canonical_dataset"),
                    "path": str(dataset.ref.json_path.relative_to(REPO_ROOT)),
                }
            )
        return items


def _index_entries(dataset: Dataset, entries: List[Dict[str, Any]]) -> None:
    dataset.entries = entries
    for entry in entries:
        entry_id = entry.get("id")
        if entry_id:
            dataset.by_id[entry_id] = entry
        for value in _headword_keys(entry):
            dataset.by_fold.setdefault(fold(value), []).append(entry)
            dataset.by_loose.setdefault(fold_loose(value), []).append(entry)


def present_entry(
    dataset: Dataset,
    entry: Dict[str, Any],
    base_url: Optional[str] = None,
) -> Dict[str, Any]:
    origin = (base_url or API_ORIGIN).rstrip("/")
    payload = copy.deepcopy(entry)
    audio_items = payload.get("audio") or []
    for item in audio_items:
        relative = (item.get("file") or "").replace("\\", "/")
        filename = Path(relative).name
        if filename:
            item["url"] = f"{origin}/v1/{dataset.ref.key}/audio/{filename}"
    return payload


def lookup_headword(dataset: Dataset, headword: str) -> Tuple[str, List[Dict[str, Any]]]:
    needle = unicodedata.normalize("NFC", headword).strip()
    if not needle:
        return "exact", []

    folded = fold(needle)
    loose = fold_loose(needle)

    exact = [
        entry
        for entry in dataset.by_fold.get(folded, [])
        if unicodedata.normalize("NFC", (entry.get("headword") or "")).strip() == needle
    ]
    if exact:
        return "exact", _unique(exact)

    folded_hits = dataset.by_fold.get(folded) or []
    if folded_hits:
        return "normalized", _unique(folded_hits)

    if len(loose) > 1:
        loose_hits = dataset.by_loose.get(loose) or []
        if loose_hits:
            return "diacritic-insensitive", _unique(loose_hits)
    return "exact", []


def search_entries(dataset: Dataset, query: str) -> List[Dict[str, Any]]:
    needle = fold(query)
    if not needle:
        return list(dataset.entries)

    scored: List[Tuple[int, int, Dict[str, Any]]] = []
    for index, entry in enumerate(dataset.entries):
        headword = fold(entry.get("headword") or "")
        translation_pt = fold(entry.get("translation_pt") or "")
        translation_en = fold(entry.get("translation_en") or "")
        entry_id = fold(entry.get("id") or "")
        if needle == headword:
            rank = 0
        elif headword.startswith(needle):
            rank = 1
        elif needle in headword:
            rank = 2
        elif needle in translation_pt or needle in translation_en:
            rank = 3
        elif needle in entry_id:
            rank = 4
        else:
            continue
        scored.append((rank, index, entry))
    scored.sort(key=lambda item: (item[0], item[1]))
    return [item[2] for item in scored]


def audio_file(dataset: Dataset, filename: str) -> Optional[Path]:
    name = Path(filename).name
    if name != filename or name in {".", ".."}:
        return None
    folder = dataset.ref.json_path.parent / "Audio"
    path = (folder / name).resolve()
    try:
        path.relative_to(folder.resolve())
    except ValueError:
        return None
    if path.is_file():
        return path
    return None


def _unique(entries: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    seen = set()
    unique = []
    for entry in entries:
        marker = id(entry)
        if marker in seen:
            continue
        seen.add(marker)
        unique.append(entry)
    return unique

File: api/src/store.ts
import type { DatasetRef } from "./catalog";
import { folderFromJsonPath, loadCatalog } from "./catalog";
import { fetchGithubJson } from "./github";
import { entryGraph, indexKnowledgeLinks } from "./graph";
import {
  DOCUMENT_COLLECTIONS,
  emptyKnowledge,
  loadKnowledge,
  type Knowledge,
} from "./knowledge";
import { API_ORIGIN } from "./settings";

const APOSTROPHES = /[‘’ʼ´`]/g;
const META_SKIP = new Set(["entries"]);

export function fold(text: string): string {
  const normalized = text.normalize("NFC").trim().toLowerCase().replace(APOSTROPHES, "'");
  return normalized.split(/\s+/).join(" ");
}

export function foldLoose(text: string): string {
  return fold(text)
    .normalize("NFD")
    .replace(/\p{M}/gu, "");
}

function headwordKeys(entry: Record<string, unknown>): string[] {
  const values: string[] = [String(entry.headword || "")];
  for (const variant of (entry.variants as unknown[]) || []) {
    if (typeof variant === "string") {
      values.push(variant);
    } else if (variant && typeof variant === "object") {
      const row = variant as Record<string, unknown>;
      values.push(String(row.form || row.headword || ""));
    }
  }
  return values.filter(Boolean);
}

export class Dataset {
  entries: Record<string, unknown>[] = [];
  byId: Record<string, Record<string, unknown>> = {};
  byFold: Record<string, Record<string, unknown>[]> = {};
  byLoose: Record<string, Record<string, unknown>[]> = {};
  knowledge: Knowledge = emptyKnowledge();
  knowledgeByEntryId: Record<string, Array<[string, Record<string, unknown>]>> = {};
  knowledgeByHeadword: Record<string, Array<[string, Record<string, unknown>]>> = {};

  constructor(
    readonly ref: DatasetRef,
    readonly document: Record<string, unknown>,
  ) {}

  get kind(): DatasetRef["kind"] {
    return this.ref.kind;
  }

  get missingTerm(): Record<string, unknown> {
    const fallback = {
      status: "error",
      code: "TERM_NOT_FOUND",
      message: "Translation not available in the verified database.",
    };
    const provided = this.document.missing_term;
    return provided && typeof provided === "object"
      ? { ...(provided as Record<string, unknown>) }
      : fallback;
  }

  metadata(): Record<string, unknown> {
    const payload: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(this.document)) {
      if (!META_SKIP.has(key)) {
        payload[key] = structuredClone(value);
      }
    }
    payload.dataset = this.ref.key;
    payload.kind = this.ref.kind;
    if (this.ref.canonicalKey) {
      payload.canonical_dataset = this.ref.canonicalKey;
    }
    payload.entry_count = this.entries.length;
    payload.knowledge = Object.fromEntries(
      DOCUMENT_COLLECTIONS.map((name) => [
        name,
        (this.knowledge[name] || []).length,
      ]),
    );
    return payload;
  }
}

export class Store {
  constructor(
    private readonly env: Env,
    readonly refs: Map<string, DatasetRef>,
    readonly datasets: Map<string, Dataset>,
  ) {}

  static async load(env: Env): Promise<Store> {
    return new Store(env, await loadCatalog(env), new Map());
  }

  async get(key: string): Promise<Dataset | undefined> {
    const cached = this.datasets.get(key);
    if (cached) {
      return cached;
    }
    const ref = this.refs.get(key);
    if (!ref) {
      return undefined;
    }
    const document = await fetchGithubJson(this.env, ref.jsonPath);
    if (!document) {
      return undefined;
    }
    const dataset = new Dataset(ref, document);
    if (ref.kind === "lexicon") {
      indexEntries(dataset, (document.entries as Record<string, unknown>[]) || []);
      dataset.knowledge = await loadKnowledge(
        this.env,
        ref.jsonPath,
        typeof document.language === "string" ? document.language : undefined,
      );
      indexKnowledgeLinks(dataset);
    }
    this.datasets.set(key, dataset);
    return dataset;
  }

  async require(key: string): Promise<Dataset> {
    const dataset = await this.get(key);
    if (!dataset) {
      throw new Error(key);
    }
    return dataset;
  }

  resolveKey(family: string, variety?: string | null): string {
    return variety ? `${family}/${variety}` : family;
  }

  async ensureAll(): Promise<void> {
    for (const key of this.refs.keys()) {
      try {
        await this.get(key);
      } catch {
        continue;
      }
    }
  }

  async catalog(): Promise<Record<string, unknown>[]> {
    return [...this.refs.values()]
      .sort((a, b) => a.key.localeCompare(b.key))
      .map((ref) => {
        const loaded = this.datasets.get(ref.key);
        const meta = loaded?.metadata();
        return {
          dataset: ref.key,
          kind: ref.kind,
          language: meta?.language,
          language_name: meta?.language_name || ref.languageName,
          iso_639_3: meta?.iso_639_3,
          entry_count: meta?.entry_count,
          canonical_dataset: ref.canonicalKey,
          path: ref.jsonPath,
          folder: folderFromJsonPath(ref.jsonPath),
        };
      });
  }

  async lexicons(): Promise<Dataset[]> {
    await this.ensureAll();
    return [...this.datasets.values()].filter((dataset) => dataset.kind === "lexicon");
  }
}

function indexEntries(dataset: Dataset, entries: Record<string, unknown>[]): void {
  dataset.entries = entries;
  for (const entry of entries) {
    const entryId = entry.id;
    if (typeof entryId === "string" && entryId) {
      dataset.byId[entryId] = entry;
    }
    for (const value of headwordKeys(entry)) {
      (dataset.byFold[fold(value)] ||= []).push(entry);
      (dataset.byLoose[foldLoose(value)] ||= []).push(entry);
    }
  }
}

export function presentEntry(
  dataset: Dataset,
  entry: Record<string, unknown>,
  baseUrl?: string,
): Record<string, unknown> {
  const origin = (baseUrl || API_ORIGIN).replace(/\/$/, "");
  const payload = structuredClone(entry);
  const audioItems = (payload.audio as Record<string, unknown>[]) || [];
  for (const item of audioItems) {
    const relative = String(item.file || "").replace(/\\/g, "/");
    const filename = relative.split("/").pop();
    if (filename) {
      item.url = `${origin}/v1/${dataset.ref.key}/audio/${filename}`;
    }
  }
  payload.graph = entryGraph(dataset, entry, origin);
  return payload;
}

export function lookupHeadword(
  dataset: Dataset,
  headword: string,
): ["exact" | "normalized" | "diacritic-insensitive", Record<string, unknown>[]] {
  const needle = headword.normalize("NFC").trim();
  if (!needle) {
    return ["exact", []];
  }
  const folded = fold(needle);
  const loose = foldLoose(needle);
  const exact = (dataset.byFold[folded] || []).filter(
    (entry) => String(entry.headword || "").normalize("NFC").trim() === needle,
  );
  if (exact.length) {
    return ["exact", unique(exact)];
  }
  const foldedHits = dataset.byFold[folded] || [];
  if (foldedHits.length) {
    return ["normalized", unique(foldedHits)];
  }
  if (loose.length > 1) {
    const looseHits = dataset.byLoose[loose] || [];
    if (looseHits.length) {
      return ["diacritic-insensitive", unique(looseHits)];
    }
  }
  return ["exact", []];
}

export function searchEntries(
  dataset: Dataset,
  query: string,
): Record<string, unknown>[] {
  const needle = fold(query);
  if (!needle) {
    return [...dataset.entries];
  }
  const scored: Array<[number, number, Record<string, unknown>]> = [];
  dataset.entries.forEach((entry, index) => {
    const headword = fold(String(entry.headword || ""));
    const translationPt = fold(String(entry.translation_pt || ""));
    const translationEn = fold(String(entry.translation_en || ""));
    const entryId = fold(String(entry.id || ""));
    let rank = -1;
    if (needle === headword) rank = 0;
    else if (headword.startsWith(needle)) rank = 1;
    else if (headword.includes(needle)) rank = 2;
    else if (translationPt.includes(needle) || translationEn.includes(needle)) rank = 3;
    else if (entryId.includes(needle)) rank = 4;
    if (rank >= 0) {
      scored.push([rank, index, entry]);
    }
  });
  scored.sort((a, b) => a[0] - b[0] || a[1] - b[1]);
  return scored.map((item) => item[2]);
}

export function audioRelativePath(dataset: Dataset, filename: string): string | null {
  const name = filename.split(/[\\/]/).pop() || "";
  if (!name || name !== filename || name === "." || name === "..") {
    return null;
  }
  const folder = dataset.ref.jsonPath.replace(/\/dictionary\.json$/, "");
  return `${folder}/Audio/${name}`;
}

function unique(entries: Record<string, unknown>[]): Record<string, unknown>[] {
  const seen = new Set<Record<string, unknown>>();
  const rows: Record<string, unknown>[] = [];
  for (const entry of entries) {
    if (seen.has(entry)) {
      continue;
    }
    seen.add(entry);
    rows.push(entry);
  }
  return rows;
}

let storePromise: Promise<Store> | null = null;
let storeLoadedAt = 0;

export function getStore(env: Env): Promise<Store> {
  const ttlMs = (Number(env.GITHUB_CACHE_TTL) || 300) * 1000;
  if (!storePromise || Date.now() - storeLoadedAt > ttlMs) {
    storeLoadedAt = Date.now();
    storePromise = Store.load(env);
  }
  return storePromise;
}
